File: skills/plotting/tests/test_stripchart.py
"""Tests for pavlab_stripchart (Python/matplotlib).
Run from skills/plotting/:
    .venv/bin/pytest tests/test_stripchart.py -v
"""

from __future__ import annotations

import sys
import warnings
from pathlib import Path

import numpy as np
import pytest

sys.path.insert(0, str(Path(__file__).resolve().parent.parent / "python"))

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.collections import LineCollection, PathCollection

from pavlab_stripchart import pavlab_stripchart


# ---- fixtures ---------------------------------------------------------------

@pytest.fixture
def simple():
    """20 points across two groups; all y positive."""
    rng = np.random.default_rng(0)
    groups = ["A"] * 10 + ["B"] * 10
    values = rng.uniform(1, 5, 20)
    return groups, values


@pytest.fixture
def three_groups():
    rng = np.random.default_rng(1)
    groups = ["ctrl"] * 15 + ["treat"] * 15 + ["dko"] * 15
    values = rng.uniform(0, 10, 45)
    return groups, values


# ---- 1. Basic rendering -----------------------------------------------------

def test_returns_axes(simple):
    ax = pavlab_stripchart(*simple)
    assert hasattr(ax, "collections")
    plt.close("all")


def test_writes_file(simple, tmp_path):
    out = tmp_path / "strip.png"
    pavlab_stripchart(*simple, filename=str(out))
    assert out.exists() and out.stat().st_size > 500
    plt.close("all")


def test_no_gridlines_no_top_right_spine(simple):
    ax = pavlab_stripchart(*simple)
    assert not ax.spines["top"].get_visible()
    assert not ax.spines["right"].get_visible()
    grid = ax.get_xgridlines() + ax.get_ygridlines()
    assert all(not ln.get_visible() for ln in grid)
    plt.close("all")


# ---- 2. Input validation ----------------------------------------------------

def test_length_mismatch_raises():
    with pytest.raises(ValueError, match="equal length"):
        pavlab_stripchart(["A", "B"], [1.0, 2.0, 3.0])


def test_color_length_mismatch_raises(simple):
    g, v = simple
    with pytest.raises(ValueError, match="color array length"):
        pavlab_stripchart(g, v, color=np.ones(len(v) + 1))


# ---- 3. Non-finite y dropped silently ---------------------------------------

def test_nonfinite_y_dropped():
    groups = ["A", "A", "B", "B", "B"]
    values = [1.0, np.nan, 2.0, np.inf, 3.0]
    ax = pavlab_stripchart(groups, values)
    # 3 finite points remain: (A,1), (B,2), (B,3)
    pts = ax.collections[-1]   # scatter PathCollection is the last collection
    assert isinstance(pts, PathCollection)
    assert len(pts.get_offsets()) == 3
    plt.close("all")


# ---- 4. Group order ---------------------------------------------------------

def test_default_group_order_matches_appearance():
    """Groups appear in the order they first appear in the input."""
    groups = ["B", "A", "B", "A"]
    values = [1.0, 2.0, 3.0, 4.0]
    ax = pavlab_stripchart(groups, values)
    labels = [t.get_text() for t in ax.get_xticklabels()]
    assert labels == ["B", "A"]
    plt.close("all")


def test_order_reorders_groups(three_groups):
    g, v = three_groups
    ax = pavlab_stripchart(g, v, order=["dko", "ctrl", "treat"])
    labels = [t.get_text() for t in ax.get_xticklabels()]
    assert labels == ["dko", "ctrl", "treat"]
    plt.close("all")


def test_order_missing_group_warns(simple):
    g, v = simple
    with pytest.warns(UserWarning, match="no data"):
        pavlab_stripchart(g, v, order=["A", "B", "MISSING"])
    plt.close("all")


# ---- 5. Mean / median reference lines ---------------------------------------

def test_show_mean_draws_hline_at_grand_mean():
    groups = ["A"] * 5 + ["B"] * 5
    values = np.arange(10, dtype=float)   # mean = 4.5
    ax = pavlab_stripchart(groups, values, show_mean=True)
    # Mean line is drawn before scatter → collections[0] is a LineCollection
    assert isinstance(ax.collections[0], LineCollection)
    segs = ax.collections[0].get_segments()
    assert len(segs) == 1
    assert segs[0][0][1] == pytest.approx(4.5)   # y-position = grand mean
    plt.close("all")


def test_show_median_draws_hline_at_grand_median():
    groups = ["A"] * 5 + ["B"] * 5
    values = np.arange(10, dtype=float)   # median = 4.5
    ax = pavlab_stripchart(groups, values, show_median=True)
    assert isinstance(ax.collections[0], LineCollection)
    segs = ax.collections[0].get_segments()
    assert segs[0][0][1] == pytest.approx(4.5)
    plt.close("all")


def test_show_mean_and_median_two_hlines():
    groups = ["A"] * 5 + ["B"] * 5
    values = np.arange(10, dtype=float)
    ax = pavlab_stripchart(groups, values, show_mean=True, show_median=True)
    line_colls = [c for c in ax.collections if isinstance(c, LineCollection)]
    assert len(line_colls) == 2
    plt.close("all")


def test_no_mean_without_flag(simple):
    ax = pavlab_stripchart(*simple, show_mean=False)
    line_colls = [c for c in ax.collections if isinstance(c, LineCollection)]
    assert len(line_colls) == 0
    plt.close("all")


# ---- 6. Log transform -------------------------------------------------------

def test_log_y_appends_label_suffix(simple):
    g, v = simple
    ax = pavlab_stripchart(g, v, log_y="log2", ylabel="Expression")
    assert "(log₂)" in ax.get_ylabel()
    plt.close("all")


def test_log_y_transforms_values():
    groups = ["A"] * 3
    values = np.array([0.0, 1.0, 3.0])   # pseudocount=1 → log2([1,2,4])=[0,1,2]
    ax = pavlab_stripchart(groups, values, log_y="log2", pseudocount=1.0,
                           origin_zero=False)
    pts = ax.collections[-1]
    y_plotted = sorted(pts.get_offsets()[:, 1].tolist())
    np.testing.assert_allclose(y_plotted, [0.0, 1.0, 2.0], atol=1e-5)
    plt.close("all")


# ---- 7. Colour modes --------------------------------------------------------

def test_uniform_black_default(simple):
    ax = pavlab_stripchart(*simple)
    pts = ax.collections[-1]
    # facecolors should all be black
    fc = pts.get_facecolors()
    assert fc.shape[1] == 4    # RGBA
    assert np.allclose(fc[:, :3], 0.0)
    plt.close("all")


def test_uniform_color_string(simple):
    g, v = simple
    ax = pavlab_stripchart(g, v, color="#2563eb")
    pts = ax.collections[-1]
    fc = pts.get_facecolors()
    # All points have the same non-black color
    assert not np.allclose(fc[:, :3], 0.0)
    assert np.allclose(fc[0], fc[-1])
    plt.close("all")


def test_categorical_color_legend(simple):
    g, v = simple
    labels = ["X"] * 10 + ["Y"] * 10
    ax = pavlab_stripchart(g, v, color=labels)
    leg = ax.get_legend()
    assert leg is not None
    handles = getattr(leg, "legend_handles", None) or leg.legendHandles
    assert len(handles) == 2
    plt.close("all")


def test_numeric_color_colorbar(simple):
    g, v = simple
    scores = np.linspace(0, 1, len(v))
    ax = pavlab_stripchart(g, v, color=scores, color_label="score")
    assert len(ax.figure.axes) == 2   # main axes + colorbar
    plt.close("all")


# ---- 8. origin_zero ---------------------------------------------------------

def test_origin_zero_pins_y_to_0(simple):
    ax = pavlab_stripchart(*simple, origin_zero=True)  # all y > 0
    assert ax.get_ylim()[0] == pytest.approx(0.0)
    plt.close("all")


def test_origin_zero_false_skips(simple):
    ax = pavlab_stripchart(*simple, origin_zero=False)
    assert ax.get_ylim()[0] > 0.0
    plt.close("all")


# ---- 9. Label size ----------------------------------------------------------

def test_label_size_large_bigger_than_small(simple):
    ax_s = pavlab_stripchart(*simple, ylabel="Y", label_size="small")
    ax_l = pavlab_stripchart(*simple, ylabel="Y", label_size="large")
    assert ax_l.get_yaxis().get_label().get_size() > \
           ax_s.get_yaxis().get_label().get_size()
    plt.close("all")


# ---- 10. ax= kwarg ----------------------------------------------------------

def test_ax_kwarg_reuses_figure(simple):
    fig, axes = plt.subplots(1, 2)
    pavlab_stripchart(*simple, ax=axes[0])
    pavlab_stripchart(*simple, ax=axes[1])
    assert len(fig.axes) == 2
    plt.close("all")

File: skills/plotting/tests/test_stripchart.R
#!/usr/bin/env Rscript
# Tests for pavlab_stripchart (R/ggplot2 backend).
# Run from skills/plotting/:
#   Rscript tests/test_stripchart.R

if (!requireNamespace("ggplot2", quietly = TRUE))
  stop("ggplot2 required: install.packages('ggplot2')")

source("R/palettes.R")
source("R/pavlab_scatter.R")
source("R/pavlab_stripchart.R")

cat("test_stripchart.R\n")
cat("=================\n")

# ---- Helper -----------------------------------------------------------------

.catch_warning <- function(expr, pattern = ".*") {
  msg <- NULL
  withCallingHandlers(expr,
    warning = function(w) {
      if (is.null(msg) && grepl(pattern, conditionMessage(w)))
        msg <<- conditionMessage(w)
      invokeRestart("muffleWarning")
    }
  )
  msg
}

# ---- Fixtures ---------------------------------------------------------------

set.seed(0)
groups <- rep(c("A", "B"), each = 10)
values <- runif(20, 1, 5)

set.seed(1)
g3 <- rep(c("ctrl", "treat", "dko"), each = 15)
v3 <- runif(45, 0, 10)

tmpdir <- tempfile("test_stripchart_")
dir.create(tmpdir)
on.exit(unlink(tmpdir, recursive = TRUE), add = TRUE)

# ---- 1. Basic rendering -----------------------------------------------------

cat("1. basic call returns gg object and writes PNG...")
p <- pavlab_stripchart(groups, values,
                       filename = file.path(tmpdir, "basic.png"))
stopifnot(inherits(p, "gg"))
stopifnot(file.exists(file.path(tmpdir, "basic.png")))
stopifnot(file.size(file.path(tmpdir, "basic.png")) > 1000)
cat(" ok\n")

cat("2. theme_classic applied: no grid, top/right off...")
built <- ggplot2::ggplot_build(p)
stopifnot(!is.null(built))
cat(" ok\n")

# ---- 2. Input validation ----------------------------------------------------

cat("3. unequal-length x/y throws error...")
err <- tryCatch(
  pavlab_stripchart(c("A", "B"), c(1, 2, 3)),
  error = function(e) conditionMessage(e)
)
stopifnot(grepl("equal length", err))
cat(" ok\n")

# ---- 3. Non-finite y dropped ------------------------------------------------

cat("4. non-finite y values dropped silently...")
g_na <- c("A", "A", "B", "B", "B")
v_na <- c(1.0, NA, 2.0, Inf, 3.0)
p_na <- pavlab_stripchart(g_na, v_na)
stopifnot(inherits(p_na, "gg"))
# 3 finite rows remain
stopifnot(nrow(p_na$data) == 3)
cat(" ok\n")

# ---- 4. Group ordering ------------------------------------------------------

cat("5. default order follows first-appearance of groups...")
p_ord <- pavlab_stripchart(c("B", "A", "B", "A"), c(1, 2, 3, 4))
lvls <- levels(p_ord$data$x_grp)
stopifnot(lvls[1] == "B")
stopifnot(lvls[2] == "A")
cat(" ok\n")

cat("6. order= parameter reorders groups...")
p_reord <- pavlab_stripchart(g3, v3, order = c("dko", "ctrl", "treat"))
lvls_reord <- levels(p_reord$data$x_grp)
stopifnot(lvls_reord[1] == "dko")
stopifnot(lvls_reord[2] == "ctrl")
stopifnot(lvls_reord[3] == "treat")
cat(" ok\n")

# ---- 5. Mean / median reference lines ---------------------------------------

cat("7. show_mean=TRUE adds a geom_hline at grand mean...")
grand_mean <- mean(values)
p_mean <- pavlab_stripchart(groups, values, show_mean = TRUE)
# Reference lines are added before geom_jitter, so layers[[1]] is geom_hline
has_hline <- inherits(p_mean$layers[[1]]$geom, "GeomHline")
stopifnot(has_hline)
hline_y <- p_mean$layers[[1]]$data$yintercept
stopifnot(abs(hline_y - grand_mean) < 1e-9)
cat(" ok\n")

cat("8. show_median=TRUE adds a geom_hline at grand median...")
grand_median <- median(values)
p_med <- pavlab_stripchart(groups, values, show_median = TRUE)
stopifnot(inherits(p_med$layers[[1]]$geom, "GeomHline"))
med_y <- p_med$layers[[1]]$data$yintercept
stopifnot(abs(med_y - grand_median) < 1e-9)
cat(" ok\n")

cat("9. show_mean + show_median → two geom_hline layers...")
p_both <- pavlab_stripchart(groups, values, show_mean = TRUE, show_median = TRUE)
hline_layers <- Filter(function(l) inherits(l$geom, "GeomHline"), p_both$layers)
stopifnot(length(hline_layers) == 2)
cat(" ok\n")

cat("10. no mean line by default...")
p_nomean <- pavlab_stripchart(groups, values)
hline_layers_no <- Filter(function(l) inherits(l$geom, "GeomHline"), p_nomean$layers)
stopifnot(length(hline_layers_no) == 0)
cat(" ok\n")

# ---- 6. Log transform -------------------------------------------------------

cat("11. log_y appends suffix to y-axis label...")
p_log <- pavlab_stripchart(groups, values, log_y = "log2", ylabel = "Expression")
stopifnot(grepl("log", p_log$labels$y))
cat(" ok\n")

cat("12. log_y transforms values in plot data...")
# values all > 0, pseudocount=1: log2(v+1)
p_logval <- pavlab_stripchart(groups, values, log_y = "log2", pseudocount = 1.0)
expected_y <- log2(values + 1.0)
actual_y   <- sort(p_logval$data$y_val)
stopifnot(all(abs(actual_y - sort(expected_y)) < 1e-9))
cat(" ok\n")

# ---- 7. Colour modes --------------------------------------------------------

cat("13. color=NULL → uniform black points...")
p_blk <- pavlab_stripchart(groups, values)
# No color_col column in data → uniform color
stopifnot(!"color_col" %in% names(p_blk$data))
cat(" ok\n")

cat("14. categorical color_col column added for char vector...")
strat <- rep(c("X", "Y"), 10)
p_cat <- pavlab_stripchart(groups, values, color = strat)
stopifnot("color_col" %in% names(p_cat$data))
built_cat <- ggplot2::ggplot_build(p_cat)
stopifnot(!is.null(built_cat))
cat(" ok\n")

cat("15. numeric color_col column added for float vector...")
scores <- runif(20)
p_num <- pavlab_stripchart(groups, values, color = scores)
stopifnot("color_col" %in% names(p_num$data))
built_num <- ggplot2::ggplot_build(p_num)
stopifnot(!is.null(built_num))
cat(" ok\n")

# ---- 8. origin_zero ---------------------------------------------------------

cat("16. origin_zero=TRUE trains y scale to include 0...")
p_oz <- pavlab_stripchart(groups, values, origin_zero = TRUE)  # all y > 0
built_oz <- ggplot2::ggplot_build(p_oz)
y_range <- built_oz$layout$panel_scales_y[[1]]$range$range
stopifnot(y_range[1] <= 0)
cat(" ok\n")

cat("17. origin_zero=FALSE leaves y range above 0...")
p_noz <- pavlab_stripchart(groups, values, origin_zero = FALSE)
built_noz <- ggplot2::ggplot_build(p_noz)
y_range_noz <- built_noz$layout$panel_scales_y[[1]]$range$range
stopifnot(y_range_noz[1] > 0.5)
cat(" ok\n")

# ---- 9. Labels --------------------------------------------------------------

cat("18. title/xlabel/ylabel appear in plot labels...")
p_labs <- pavlab_stripchart(groups, values, title = "Title",
                             xlabel = "Group", ylabel = "Value")
stopifnot(p_labs$labels$title == "Title")
stopifnot(grepl("Group", p_labs$labels$x))
stopifnot(grepl("Value", p_labs$labels$y))
cat(" ok\n")

# ---- 10. kind="swarm" -------------------------------------------------------

cat("19. kind='strip' (default) jitters the point layer...")
p_strip <- pavlab_stripchart(groups, values)  # default kind
# geom_jitter reports class "GeomPoint" — the jitter is on the layer's
# position, not the geom. Look for at least one PositionJitter layer.
positions <- vapply(p_strip$layers, function(l) class(l$position)[1], character(1))
stopifnot("PositionJitter" %in% positions)
cat(" ok\n")

if (requireNamespace("ggbeeswarm", quietly = TRUE)) {
  cat("20. kind='swarm' uses GeomBeeswarm when ggbeeswarm is available...")
  p_sw <- pavlab_stripchart(groups, values, kind = "swarm")
  geoms_sw <- vapply(p_sw$layers, function(l) class(l$geom)[1], character(1))
  stopifnot("GeomBeeswarm" %in% geoms_sw)
  cat(" ok\n")
} else {
  cat("20. kind='swarm' raises clear error when ggbeeswarm missing...")
  err <- tryCatch({
    pavlab_stripchart(groups, values, kind = "swarm"); NULL
  }, error = function(e) conditionMessage(e))
  stopifnot(!is.null(err), grepl("ggbeeswarm", err))
  cat(" ok\n")
}

cat("21. kind='violin' rejected by match.arg...")
err2 <- tryCatch({
  pavlab_stripchart(groups, values, kind = "violin"); NULL
}, error = function(e) conditionMessage(e))
stopifnot(!is.null(err2), grepl("should be one of", err2))
cat(" ok\n")

cat("\nALL TESTS PASSED\n")
